### PR TITLE
Add SourceDistPlugin for creating source packages

### PR DESCRIPTION
**Pull request by mdedetrich**
Adds the prototype source dist plugin which generates source packages for Apache

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -100,9 +100,12 @@ lazy val root = Project(id = "pekko", base = file("."))
       akkaScalaNightly,
       docs,
       serialversionRemoverPlugin))
-  .settings(Compile / headerCreate / unmanagedSources := (baseDirectory.value / "project").**("*.scala").get)
+  .settings(
+    Compile / headerCreate / unmanagedSources := (baseDirectory.value / "project").**("*.scala").get,
+    sourceDistName := "pekko")
```
Comment (pjfanning) [inline build.sbt:105]: incubator-pekko is more correct for the source distribution

Comment (mdedetrich) [reply build.sbt]: Fixed

**Review by pjfanning — APPROVED** [review]
lgtm

**Review by pjfanning — CHANGES_REQUESTED** [review]
there was a problem with v0.1.0 of the plugin - some params passed to the generator were in wrong order (version and targetDir) - which meant the output files were getting created with wrong dir and names - I've published a v0.1.1 with a fix

**Comment by mdedetrich** [comment]
@pjfanning Fixed

**Review by gmethvin — APPROVED** [review]

**Review by pjfanning — APPROVED** [review]

**Comment by pjfanning** [comment]
v0.1.2 of the sbt plugin is available now

**Comment by mdedetrich** [comment]
> v0.1.2 of the sbt plugin is available now

Will update

**Comment by spangaer** [comment]
Sorry for the late comment, but wondering whether something like this was considered:
Query Git for an archive of the branch state: https://git-scm.com/docs/git-archive

I guess the archive can receive some post processing to added certain "generated files", like the "dist stuff" or
https://github.com/apache/incubator-pekko/issues/130#issuecomment-1421339657

**Comment by pjfanning** [comment]
@spangaer since we've already written the sbt plugin, I don't see the real attraction to using this command line tool. If we hadn't written the plugin, the tool might have been an alternative.

File: project/plugins.sbt [hunk 2]
```diff
@@ -26,4 +26,5 @@ addSbtPlugin("net.bzzt" % "sbt-reproducible-builds" % "0.30")
 
 addSbtPlugin("com.github.sbt" % "sbt-ci-release" % "1.5.10")
 addSbtPlugin("com.lightbend.sbt" % "sbt-publish-rsync" % "0.2")
+addSbtPlugin("com.github.pjfanning" % "sbt-source-dist" % "0.1.2")
```
Comment (pjfanning) [inline project/plugins.sbt:29]: 0.1.3 is ready - it generates the sha256/sha512 digest files as well as the zip/tgz

**Comment by spangaer** [comment]
Well, I'd think you'll always need a plugin to do more. I just saw things like "self handling" of .gitignore files, to which I think -> just let Git handle that. Where wrapping that command would help you.

But I fully understand I'm waaaay late with that suggestion.

**Review by jrudolph — APPROVED** [review]
lgtm

I compared `git archive` output with the output of this plugin and it seems to work well-enough (though being 10-100x slower than `git archive`). The most important thing here is that you start from a pristine checkout because files are included based on a blacklist and not on a whitelist so any extra file you might have in your workspace will be included.

I would have also preferred to simply use `git archive` but since the work is already done, it seems good enough.

**Comment by mdedetrich** [comment]
> The most important thing here is that you start from a pristine checkout because files are included based on a blacklist and not on a whitelist so any extra file you might have in your workspace will be included.

We can add this as a check to sbt-source-dist so it won't work if the checkout is not pristine

> I would have also preferred to simply use git archive but since the work is already done, it seems good enough.

Its always possible to make sbt-source-dist just wrap `git archive` however arguably this is also an advantage of sbt-source-dist because its not reliant on git. Not relevant for pekko specifically but as a general Apache SBT plugin (which it may be in the future) this has theoretical relevance because not everyone uses git (then again, a good counter argument to this is to just detect if the machine has git and then proceed to use git archive).

**Comment by pjfanning** [comment]
I agree that `git archive` would be a workable solution.

A few benefits of an sbt plugin
* the artifact names are set up in the sbt scripts so that users just call `sbt sourceDistGenerate` regardless of the repo - so they don't need to customise the commands for different repos. One call gets you the zip and tgz
* the `sourceDistGenerate` creates the SHA digests automatically
* we'll probably also get it use sbt-pgp to gen the gpg signing asc files
* checking for a clean unmodified repo is definitely needed

If the plugin proves problematic, we can look again at `git archive`.

**Comment by mdedetrich** [comment]
> If the plugin proves problematic, we can look again at git archive.

Just a note on this (incase its not clear), there isn't any issue in sbt plugins just wrap existing cli's (such as `git` with `git archive`). In fact a lot of sbt plugins, even official ones do this (a good example is sbt-pgp).

**Comment by jrudolph** [comment]
Just as quick info that in my checkout, the plugin somehow got into a loop and created a 50GB distribution file when I noticed and stopped it...

**Comment by mdedetrich** [comment]
> Just as quick info that in my checkout, the plugin somehow got into a loop and created a 50GB distribution file when I noticed and stopped it...

Hmm, I am honestly thinking that in the medium term, let alone long term it would be easier to divert to git archive. @jrudolph Do you want to investigate https://github.com/pjfanning/sbt-source-dist/issues/14 ? 

@pjfanning Can you add @jrudolph as a maintainer of the repo as well?

**Comment by mdedetrich** [comment]
We can also have a look into what https://github.com/Apache/daffodil does, its another Apache project that uses sbt

**Pull request merged**